src/bacnet/basic/object/calendar.c

```diff
@@ -0,0 +1,757 @@
+}
+
+/**
+ * @brief Determines a object write-enabled flag state
+ * @param object_instance - object-instance number of the object
+ * @return  write-enabled status flag
+ */
+bool Calendar_Write_Enabled(uint32_t object_instance)
+{
+    bool value = false;
+    struct object_data *pObject;
+
+    pObject = Keylist_Data(Object_List, object_instance);
+    if (pObject) {
+        value = pObject->Write_Enabled;
+    }
+
+    return value;
+}
+
+/**
+ * @brief For a given object instance-number, sets the write-enabled flag
+ * @param object_instance - object-instance number of the object
+ */
+void Calendar_Write_Enable(uint32_t object_instance)
+{
+    struct object_data *pObject;
+
+    pObject = Keylist_Data(Object_List, object_instance);
+    if (pObject) {
+        pObject->Write_Enabled = true;
+    }
+}
+
+/**
+ * @brief For a given object instance-number, clears the write-enabled flag
+ * @param object_instance - object-instance number of the object
+ */
+void Calendar_Write_Disable(uint32_t object_instance)
+{
+    struct object_data *pObject;
+
+    pObject = Keylist_Data(Object_List, object_instance);
+    if (pObject) {
+        pObject->Write_Enabled = false;
+    }
+}
+
+/**
+ * Creates a Calendar object
+ * @param object_instance - object-instance number of the object
+ */
+bool Calendar_Create(uint32_t object_instance)
+{
+    bool status = false;
+    struct object_data *pObject = NULL;
+    int index = 0;
+
+    pObject = Keylist_Data(Object_List, object_instance);
+    if (!pObject) {
+        pObject = calloc(1, sizeof(struct object_data));
+        if (pObject) {
+            pObject->Object_Name = NULL;
+            pObject->Description = NULL;
+            pObject->Present_Value  = false;
+            pObject->Date_List = Keylist_Create();
+            pObject->Changed = false;
+            pObject->Write_Enabled = false;
+            /* add to list */
+            index = Keylist_Data_Add(Object_List, object_instance, pObject);
+            if (index >= 0) {
+                status = true;
+                Device_Inc_Database_Revision();
+            }
+        }
+    }
+
+    return status;
+}
+
+/**
+ * Deletes an Calendar object
+ * @param object_instance - object-instance number of the object
+ * @return true if the object is deleted
+ */
+bool Calendar_Delete(uint32_t object_instance)
+{
+    bool status = false;
+    struct object_data *pObject = NULL;
+
+    pObject = Keylist_Data_Delete(Object_List, object_instance);
+    if (pObject) {
+        free(pObject->Date_List);
+        free(pObject);
+        status = true;
+        Device_Inc_Database_Revision();
+    }
+
+    return status;
+}
+
+/**
+ * Deletes all the Calendars and their data
+ */
+void Calendar_Cleanup(void)
+{
+    struct object_data *pObject;
+
+    if (Object_List) {
+        do {
+            pObject = Keylist_Data_Pop(Object_List);
+            if (pObject) {
+                free(pObject->Date_List);
```

Comment:
Since pObject->Date_List is a Keylist, it will need to be freed with Keylist_Delete() after freeing all the children.